### PR TITLE
Gameroomd websocket reconnect

### PR DESCRIPTION
**Pull request by eloaverona**
This PR implements:
- Gameroomd app auth handler and xo_state_delta subscribers websocket clients attempt to reconnect if they receive a "close" message from the ws server
- Gameroomd app auth handler and xo_state_delta subscribers websocket clients drop the connect and attempt to reconnect if they get no response or no messages from the server after an specified timeout.

### CODE REVIEW

File: libsplinter/src/events/ws.rs [hunk 1]
```diff
@@ -447,11 +471,22 @@ fn do_shutdown(
 pub struct Context<T: ParseBytes<T> + 'static> {
     igniter: Igniter,
     ws: WebSocketClient<T>,
+    reconnect_count: u64,
+    last_reconnect: SystemTime,
+    wait: Duration,
+    reactor_running: Arc<AtomicBool>,
```
Comment (RyanLassigBanks) [inline libsplinter/src/events/ws.rs]: I don't think `reactor_running` is necessary since `igniter` has the same variable. If `Context` needs to know if it's reactor is running it should call a method from `igniter`.

Additionally, I wonder if it would make more sense for the igniter to check if its reactor is still running an emit an error when context or a callback call `start_ws` or `send`. I don't think that needs to be addressed in this PR, just curious what people think.

Comment (eloaverona) [reply libsplinter/src/events/ws.rs]: yeah, I can use the reactor_running in the igniter, and will make that change. But I need the Context to know that the reactor is still running, otherwise, if the gameroom daemon is stopped while the context is waiting before attempt to reconnect to a ws server, the shutdown would be delayed until the wait time is over
https://github.com/Cargill/splinter/blob/cc31433dfb446221104452e0d0437a7e9779fdc1/libsplinter/src/events/ws.rs#L541-L554

Comment (eloaverona) [reply libsplinter/src/events/ws.rs]: updated

File: libsplinter/src/events/reactor.rs [hunk 2]
```diff
@@ -29,11 +33,15 @@ use crate::events::{ReactorError, WebSocketError};
 pub struct Reactor {
     sender: Sender<ReactorMessage>,
     thread_handle: thread::JoinHandle<()>,
+    running: Arc<AtomicBool>,
```
Comment (peterschwarz) [inline libsplinter/src/events/reactor.rs]: This change could be in a separate commit, with more explanation.

File: libsplinter/src/rest_api/events.rs [hunk 3]
```diff
@@ -164,6 +165,10 @@ impl<T: Serialize + Debug + 'static> Actor for EventDealerWebSocket<T> {
         if let Some(recv) = self.recv.take() {
             debug!("Starting Event Websocket");
             ctx.add_stream(recv);
+            ctx.run_interval(Duration::from_secs(30), move |_, ctx| {
```
Comment (peterschwarz) [inline libsplinter/src/rest_api/events.rs]: magic number

File: libsplinter/src/events/ws.rs [hunk 4]
```diff
@@ -338,6 +414,9 @@ impl<T: ParseBytes<T> + 'static> WebSocketClient<T> {
                                                 err
                                             );
                                         }
+                                        if let Err(err) = context.try_reconnect() {
```
Comment (peterschwarz) [inline libsplinter/src/events/ws.rs]: Should it try in this case?  Isn't this an explicit close?

Comment (eloaverona) [reply libsplinter/src/events/ws.rs]: Yes, this is called if the ws gets a close message from the server. I talked with @agunde406 about this, I guess the expectation was that even if the splinter daemon gracefully closes the connection the gameroomd should try to re-establish with the expectation that the server will return soon? The gameroom does work properly without those connections.

File: libsplinter/src/events/reactor.rs (hunk 2 above)
Comment (eloaverona) [reply libsplinter/src/events/reactor.rs]: updated

File: libsplinter/src/rest_api/events.rs (hunk 3 above)
Comment (eloaverona) [reply libsplinter/src/rest_api/events.rs]: updated

File: examples/gameroom/daemon/src/authorization_handler/mod.rs [hunk 5]
```diff
@@ -68,14 +68,25 @@ pub fn run(
         },
     );
 
+    ws.set_reconnect(true);
+    ws.set_reconnect_limit(10);
```
Comment (peterschwarz) [inline examples/gameroom/daemon/src/authorization_handler/mod.rs]: Should these be using a configured value (as in the previous commit configures)?

Comment (eloaverona) [reply examples/gameroom/daemon/src/authorization_handler/mod.rs]: Do you mean just using the defaults establish in the server side?

Comment (eloaverona) [reply examples/gameroom/daemon/src/authorization_handler/mod.rs]: addressed
